osquery/tables/system/windows/drivers.cpp

```diff
@@ -168,46 +173,43 @@ Status getDeviceProperty(const device_infoset_t& infoset,
   return Status::success();
 }
 
-std::string getDriverImagePath(const std::string& service_key) {
-  QueryData results;
-  queryMultipleRegistryPaths({service_key + kRegSep + "ImagePath"}, results);
-
-  if (results.empty()) {
-    return "";
+Status getDriverImagePath(const std::string& svc_name, std::string& result) {
+  HKEY hkey;
+  const auto image_path_value = L"ImagePath";
+  const auto svc_key = "SYSTEM\\CurrentControlSet\\Services\\" + svc_name;
+  auto ret = RegOpenKeyExW(
+      HKEY_LOCAL_MACHINE, stringToWstring(svc_key).c_str(), 0, KEY_READ, &hkey);
+  if (ret != ERROR_SUCCESS) {
+    return Status(ret, "Failed to open registry handle");
   }
 
-  auto data_it = results[0].find("data");
-  if (data_it == results[0].end()) {
-    return "";
+  reg_handle_t registry_handle(hkey, close_reg_handle);
+  DWORD buff_size;
+  ret = RegGetValueW(hkey,
+                     nullptr,
+                     image_path_value,
+                     RRF_RT_REG_SZ,
+                     nullptr,
+                     nullptr,
+                     &buff_size);
+  if (ret != ERROR_SUCCESS) {
+    return Status(ret, "Failed to query registry value(length)");
   }
 
-  auto path = data_it->second;
-  if (path.empty()) {
-    return "";
+  auto buff = std::make_unique<BYTE[]>(buff_size);
+  ret = RegGetValueW(hkey,
+                     nullptr,
+                     image_path_value,
+                     RRF_RT_REG_SZ,
+                     nullptr,
+                     buff.get(),
+                     &buff_size);
+  if (ret != ERROR_SUCCESS) {
+    return Status(ret, "Failed to query registry value");
   }
 
-  return kNormalizeImage(path);
-}
-
-Status genServiceKeyMap(
-    std::map<std::string, std::string>& services_image_map) {
-  QueryData results;
-  queryMultipleRegistryPaths({kServiceKeyPath + '%' + kRegSep + "ImagePath"},
-                             results);
-
-  // Something went wrong
-  if (results.empty()) {
-    return Status::failure("Failed to retrieve services image path cache");
-  }
-
-  for (auto& row : results) {
-    auto key_it = row.find("key");
-    auto data_it = row.find("data");
-    if (key_it == row.end() || data_it == row.end()) {
-      continue;
-    }
-    services_image_map[key_it->second] = kNormalizeImage(data_it->second);
-  }
+  auto path = wstringToString(reinterpret_cast<wchar_t*>(buff.get()));
```

Comment:
I would avoid using `reinterpret_cast` as much as possible and simply initialize `buff` above as `std::make_unique<wchar_t[]>(buff_size / sizeof(wchar_t))`, so that it's already of the correct type.
There's no need to later convert to a `BYTE*` in the `RegGetValueW` call.

Something similar, but with `std::wstring` is suggested here: https://docs.microsoft.com/en-us/archive/msdn-magazine/2017/may/c-use-modern-c-to-access-the-windows-registry

Reply:
Although `std::wstring` is straightforward wrapper and have support of dynamic memory reallocation , I choose to stick with `unique_ptr` to allocate memory, due to i know the exact memory size of the registry value from the first call to `RegGetValueW`.